### PR TITLE
Feat(tsql): improve support for transaction statements

### PR DESCRIPTION
**Pull request by dmoore247**
Enrich syntax parsing for BEGIN, COMMIT and ROLLBACK Transactions.

### CODE REVIEW

File: sqlglot/dialects/tsql.py [hunk 1]
```diff
@@ -281,8 +282,16 @@ class Tokenizer(tokens.Tokenizer):
         QUOTES = ["'", '"']
         HEX_STRINGS = [("0x", ""), ("0X", "")]
 
+        COMMAND_PREFIX_TOKENS = {TokenType.SEMICOLON}
+
         KEYWORDS = {
             **tokens.Tokenizer.KEYWORDS,
+            "BEGIN": TokenType.COMMAND,
+            "BEGIN TRANSACTION": TokenType.BEGIN,
+            "COMMIT": TokenType.COMMIT,
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:291]: these are unnecessary, why did you add them?

File: sqlglot/dialects/tsql.py [hunk 2]
```diff
@@ -358,13 +370,75 @@ class Parser(parser.Parser):
         STATEMENT_PARSERS = {
             **parser.Parser.STATEMENT_PARSERS,
             TokenType.END: lambda self: self._parse_command(),
+            TokenType.ALIAS: lambda self: self._parse_alias(),
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:373]: why did you add this

File: sqlglot/dialects/tsql.py [hunk 3]
```diff
@@ -358,13 +370,75 @@ class Parser(parser.Parser):
...
+
+            Returns:
+                exp.Commit | exp.Rollback: _description_
+            """
+            rollback = False
+            if self._prev.token_type == TokenType.ROLLBACK:
+                rollback = True
+
+            transaction = None
+            if self._match_texts({"TRAN", "TRANSACTION"}):
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:401]: transaction = self._match_texts and self._prev.text

File: sqlglot/dialects/tsql.py [hunk 4]
```diff
@@ -358,13 +370,75 @@ class Parser(parser.Parser):
...
+                transaction = self._prev.text
+            txn_name = self._parse_id_var()
+
+            durability = None
+            if self._match_text_seq("WITH", "(", "DELAYED_DURABILITY", "="):
```
Comment (tobymao) [inline sqlglot/dialects/tsql.py:406]: don't do this, use real matching

self.match(TokenType.With):
self .match_l_paren() etc

File: sqlglot/expressions.py [hunk 5]
```diff
@@ -3960,7 +3966,7 @@ def else_(self, condition: ExpOrStr, copy: bool = True, **opts) -> Case:
 
 
 class Cast(Func):
-    arg_types = {"this": True, "to": True, "format": False}
+    arg_types = {"this": True, "to": False, "format": False}
```
Comment (tobymao) [inline sqlglot/expressions.py:3969]: why did you change this, this is probably not correct

File: sqlglot/parser.py [hunk 6]
```diff
@@ -1205,7 +1205,8 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
             extend_props(self._parse_properties())
 
             self._match(TokenType.ALIAS)
-            begin = self._match(TokenType.BEGIN)
+            begin = self._match_text_seq("BEGIN")
+            #            begin = self._match(TokenType.BEGIN)
```
Comment (tobymao) [inline sqlglot/parser.py:1209]: why did you change this

File: sqlglot/dialects/tsql.py (hunk 3 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py:401]: Why do we need `_prev.text` in the first place here? If I understand this correctly, we can always just output `TRANSACTION` instead of the abbreviated form, so we can also use a flag to detect this:

```python
transaction = self._match_texts(...)
```

File: sqlglot/dialects/tsql.py [hunk 7]
```diff
@@ -539,3 +615,35 @@ def returning_sql(self, expression: exp.Returning) -> str:
...
+        def _durability_sql(self, expression) -> str:
+            durability = expression.args.get("durability")
+            durability_sql = ""
+            if durability is not None:
+                if durability:
+                    durability = "ON"
+                else:
+                    durability = "OFF"
+                durability_sql = f" WITH (DELAYED_DURABILITY = {durability})"
+            return durability_sql
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:635]:
```suggestion
            if durability is not None:
                return f" WITH (DELAYED_DURABILITY = {'ON' if durability else 'OFF'})"
            return ""
```

Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py:635]: Let's just get rid of this whole method and move the suggested logic in `commit_sql` instead, since that's where `durability` is used.

File: sqlglot/dialects/tsql.py [hunk 8]
```diff
@@ -539,3 +615,35 @@ def returning_sql(self, expression: exp.Returning) -> str:
...
+            this = self.sql(expression, "this")
+            this = f" {this}" if this else ""
+            transaction = expression.args.get("transaction")
+            transaction = f" {transaction}" if transaction else ""
+            return f"COMMIT{transaction}{this}{self._durability_sql(expression)}"
+
+        def rollback_sql(self, expression: exp.Rollback) -> str:
+            this = self.sql(expression, "this")
+            this = f" {this}" if this else ""
+            transaction = expression.args.get("transaction")
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:647]:
```suggestion
            transaction = self.sql(expression, "transaction")
```

File: sqlglot/dialects/tsql.py [hunk 9]
```diff
@@ -539,3 +615,35 @@ def returning_sql(self, expression: exp.Returning) -> str:
             into = self.sql(expression, "into")
             into = self.seg(f"INTO {into}") if into else ""
             return f"{self.seg('OUTPUT')} {self.expressions(expression, flat=True)}{into}"
+
+        def transaction_sql(self, expression: exp.Transaction) -> str:
+            this = self.sql(expression, "this")
+            this = f" {this}" if this else ""
+            mark = expression.args.get("mark")
+            mark = f" WITH MARK {mark}" if mark else ""
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:623]: This is needed otherwise you implicitly use the default dialect for `mark` when interpolating it.

```suggestion
            mark = self.sql(expression, "mark")
            mark = f" WITH MARK {mark}" if mark else ""
```

File: sqlglot/expressions.py [hunk 10]
```diff
@@ -3472,15 +3472,21 @@ class Command(Expression):
...
 
 
 class Commit(Expression):
-    arg_types = {"chain": False}
+    arg_types = {
+        "chain": False,
+        "this": False,
+        "transaction": False,
+        "durability": False,
+        "modes": False,
```
Comment (GeorgeSittas) [inline sqlglot/expressions.py:3484]: Doesn't seem like `modes` is used anywhere. Let's delete it if it isn't needed.

File: tests/dialects/test_tsql.py [hunk 11]
```diff
@@ -389,7 +389,62 @@ def test_types_bin(self):
             },
         )
 
+    def test_ddl(self):
+        sql = """CREATE TABLE #mytemp (a INTEGER,  b CHAR(2), c TIME(4), d FLOAT(24));"""
+        self.validate_all(
+            " ".join(sql.split()),
```
Comment (GeorgeSittas) [inline tests/dialects/test_tsql.py:395]: This doesn't seem correct. You can do:

```python
self.validate_identity(
    "CREATE TABLE #mytemp (a INTEGER, b CHAR(2), c TIME(4), d FLOAT(24))"
)
```

File: tests/dialects/test_tsql.py [hunk 12]
```diff
@@ -389,7 +389,62 @@ def test_types_bin(self):
...
+        # self.validate_identity("BEGIN TRAN")
+        self.validate_identity("BEGIN TRANSACTION transaction_name")
+        self.validate_identity("BEGIN TRANSACTION @tran_name_variable")
+        self.validate_identity("BEGIN TRANSACTION transaction_name WITH MARK 'description'")
+
+    def test_commit(self):
+        # COMMIT [ { TRAN | TRANSACTION }  [ transaction_name | @tran_name_variable ] ] [ WITH ( DELAYED_DURABILITY = { OFF | ON } ) ] [ ; ]
+
+        self.validate_identity("COMMIT")
+        self.validate_all("COMMIT TRAN")
```
Comment (GeorgeSittas) [inline tests/dialects/test_tsql.py:417]: This doesn't seem correct as well. Assuming we make the change I suggested in the parser, this should become:

```python
self.validate_all("COMMIT TRAN", write={"tsql": "COMMIT TRANSACTION"})
```

File: tests/dialects/test_tsql.py [hunk 13]
```diff
@@ -389,7 +389,62 @@ def test_types_bin(self):
...
+        )
+
+    def test_rollback(self):
+        # Applies to SQL Server and Azure SQL Database
+        # ROLLBACK { TRAN | TRANSACTION }
+        #     [ transaction_name | @tran_name_variable
+        #     | savepoint_name | @savepoint_variable ]
+        # [ ; ]
+        self.validate_identity("ROLLBACK")
+        self.validate_all("ROLLBACK TRAN")
```
Comment (GeorgeSittas) [inline tests/dialects/test_tsql.py:436]: Ditto

File: tests/dialects/test_tsql.py [hunk 14]
```diff
@@ -389,7 +389,62 @@ def test_types_bin(self):
...
+        self.validate_identity("ROLLBACK TRANSACTION transaction_name")
+        self.validate_identity("ROLLBACK TRANSACTION @tran_name_variable")
+
     def test_udf(self):
+        self.validate_identity("BEGIN")
+        self.validate_identity("END")
+        self.validate_identity("SET XACT_ABORT ON")
+        self.validate_identity(
+            "DECLARE @DWH_DateCreated DATETIME = CONVERT(DATETIME, getdate(), 104)"
+        )
```
Comment (GeorgeSittas) [inline tests/dialects/test_tsql.py:447]: These don't look like they belong in `test_udf`. Let's move them in `test_tsql`.

File: tests/dialects/test_tsql.py [hunk 15]
```diff
@@ -446,6 +501,7 @@ def test_udf(self):
             pretty=True,
         )
 
+    def test_fullproc(self):
```
Comment (GeorgeSittas) [inline tests/dialects/test_tsql.py:504]: Why was this moved into a different testing method?

File: sqlglot/expressions.py (hunk 10 above)
Comment (dmoore247) [reply sqlglot/expressions.py:3484]: modes is used by Presto

File: sqlglot/dialects/tsql.py (hunk 1 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:291]: "BEGIN" and "BEGIN TRANSACTION" mappings are required to disambiguate the flow control BEGIN..END and the transactional uses of BEGIN.

Comment (dmoore247) [reply sqlglot/dialects/tsql.py:291]: COMMIT and ROLLBACK can be removed.

File: sqlglot/dialects/tsql.py (hunk 2 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:373]: I put this in to try to work on CREATE PROCEDURE refinement. I'll remove for now.

File: sqlglot/dialects/tsql.py [hunk 16]
```diff
@@ -358,13 +370,75 @@ class Parser(parser.Parser):
...
+                ROLLBACK { TRAN | TRANSACTION }
+                    [ transaction_name | @tran_name_variable
+                    | savepoint_name | @savepoint_variable ]
+                    [ ; ]
+
+            Returns:
+                exp.Commit | exp.Rollback: _description_
+            """
+            rollback = False
+            if self._prev.token_type == TokenType.ROLLBACK:
```
[the remark this answers is not in the document]
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:397]: Looks good, accept

File: sqlglot/dialects/tsql.py (hunk 4 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:406]: I could use a little help on this one, it falls apart on me as soon as I add self.match_l_paren().

File: sqlglot/expressions.py (hunk 5 above)
Comment (dmoore247) [reply sqlglot/expressions.py:3969]: I had a CONVERT test case out of place and fell into changing this. I'll revert.

File: sqlglot/dialects/tsql.py (hunk 4 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py:406]: You could also write this as:

```python
if self._match_pair(TokenType.WITH, TokenType.L_PAREN):
    self._match_text_seq("DELAYED_DURABILITY")
    self._match(TokenType.EQ)
    ...
```

File: sqlglot/dialects/tsql.py (hunk 7 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:635]: Done.

File: sqlglot/dialects/tsql.py (hunk 3 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:401]: Done.

File: sqlglot/dialects/tsql.py (hunk 4 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:406]: That works.

File: sqlglot/dialects/tsql.py (hunk 8 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:647]: Ok, replaced the two instances of `expression.args.get("transaction")`

File: sqlglot/dialects/tsql.py (hunk 9 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:623]: Done

File: sqlglot/expressions.py (hunk 10 above)
Comment (dmoore247) [reply sqlglot/expressions.py:3484]: presto.py: 348
```python
def transaction_sql(self, expression: exp.Transaction) -> str:
            modes = expression.args.get("modes")
            modes = f" {', '.join(modes)}" if modes else ""
            return f"START TRANSACTION{modes}"
```

File: tests/dialects/test_tsql.py (hunk 11 above)
Comment (dmoore247) [reply tests/dialects/test_tsql.py:395]: I had another test for spark/databricks, but I removed that from this PR. I'll clean this for now.

Comment (dmoore247) [reply tests/dialects/test_tsql.py:395]: Also TIME(4) gets mapped to TIMESTAMP(4)

Comment (dmoore247) [reply tests/dialects/test_tsql.py:395]: I removed the weird join.split() which was used for the other match.

**Comment by GeorgeSittas** [comment]
@dmoore247 let us know when you push the changes so we can re-review. Thanks for the quick iterations here.

File: tests/dialects/test_tsql.py (hunk 12 above)
Comment (dmoore247) [reply tests/dialects/test_tsql.py]: Done.

File: tests/dialects/test_tsql.py (hunk 13 above)
Comment (dmoore247) [reply tests/dialects/test_tsql.py]: Done.

File: sqlglot/expressions.py (hunk 10 above)
Comment (dmoore247) [reply sqlglot/expressions.py:3483]: You're right about the Commit(), fixing.

File: tests/dialects/test_tsql.py (hunk 15 above)
Comment (GeorgeSittas) [reply tests/dialects/test_tsql.py:503]: I think I'd just move `BEGIN`, `END`, `SET XACT_ABORT ON` to `test_tsql`. I'm not sure if the `DECLARE` one is udf-related, but if it isn't I'd move that there too. We can keep `test_fullproc` as is, maybe rename to a non-abbreviated form like `test_procedure_definition`?

File: sqlglot/dialects/tsql.py [hunk 17]
```diff
@@ -283,6 +284,8 @@ class Tokenizer(tokens.Tokenizer):
 
         KEYWORDS = {
             **tokens.Tokenizer.KEYWORDS,
+            "BEGIN": TokenType.COMMAND,
+            "SET": TokenType.SET,
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:288]: This and `"WITH": TokenType.WITH` are redundant, these mappings are already in the base tokenizer.

File: sqlglot/dialects/tsql.py [hunk 18]
```diff
@@ -365,6 +371,63 @@ class Parser(parser.Parser):
...
+                    [ WITH ( DELAYED_DURABILITY = { OFF | ON } ) ]
+                    [ ; ]
+
+                ROLLBACK { TRAN | TRANSACTION }
+                    [ transaction_name | @tran_name_variable
+                    | savepoint_name | @savepoint_variable ]
+                    [ ; ]
+
+            Returns:
+                exp.Commit | exp.Rollback: _description_
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:387]: Let's get rid of this part, doesn't add much information

File: sqlglot/dialects/tsql.py [hunk 19]
```diff
@@ -365,6 +371,63 @@ class Parser(parser.Parser):
...
+        def _parse_transaction(self) -> exp.Transaction:
+            """Syntax:
+                BEGIN { TRAN | TRANSACTION }
+                [ { transaction_name | @tran_name_variable }
+                [ WITH MARK [ 'description' ] ]
+                ]
+                [ ; ]
+
+            Returns:
+                exp.Transaction: _description_
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:420]: Ditto

File: sqlglot/dialects/tsql.py [hunk 20]
```diff
@@ -365,6 +371,63 @@ class Parser(parser.Parser):
...
+            """
+            mark = None
+
+            txn_name = self._parse_id_var()
+
+            if self._match_text_seq("WITH", "MARK"):
+                mark = self._parse_string()
+
+            return self.expression(exp.Transaction, this=txn_name, mark=mark)
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:429]: Nit: can also write this a bit more compactly

```python
transaction = self.expression(exp.Transaction, this=self._parse_id_var())
if self._match_text_seq("WITH", "MARK"):
    transaction.set("mark", self._parse_string())

return self.expression(exp.Transaction
```

**Review by GeorgeSittas — COMMENTED** [review]
Ok I think there are a few comments left and this should be good to go.

File: sqlglot/parser.py (hunk 6 above)
Comment (dmoore247) [reply sqlglot/parser.py]: TokenType.BEGIN refers to Begin transaction where in `CREATE PROCEDURE` here in the create method we want BEGIN / END as in control flow ?

File: tests/dialects/test_tsql.py (hunk 14 above)
Comment (dmoore247) [reply tests/dialects/test_tsql.py]: Moved.

File: sqlglot/dialects/tsql.py (hunk 20 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Would your return be:
```python
return transaction
```

Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py]: That's correct, made a typo. Edited.

File: sqlglot/dialects/tsql.py (hunk 18 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:387]: Ok

File: sqlglot/dialects/tsql.py (hunk 19 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:420]: Ok

File: sqlglot/dialects/tsql.py (hunk 1 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Removed these tokens.

File: sqlglot/dialects/tsql.py (hunk 17 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Removed.

File: sqlglot/dialects/tsql.py (hunk 18 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Removed

File: sqlglot/dialects/tsql.py [hunk 21]
```diff
@@ -539,3 +597,31 @@ def returning_sql(self, expression: exp.Returning) -> str:
...
+            this = self.sql(expression, "this")
+            if this in ["TRAN", "TRANSACTION"]:
+                mark = self.sql(expression, "mark")
+                mark = f" WITH MARK {mark}" if mark else ""
+                return f"BEGIN TRANSACTION{mark}"
+            else:
+                this = f" {this}" if this else ""
+                mark = self.sql(expression, "mark")
+                mark = f" WITH MARK {mark}" if mark else ""
+                return f"BEGIN TRANSACTION{this}{mark}"
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:609]:
```suggestion
        def transaction_sql(self, expression: exp.Transaction) -> str:
            this = self.sql(expression, "this")
            this = f" {this}" if this else ""
            mark = self.sql(expression, "mark")
            mark = f" WITH MARK {mark}" if mark else ""
            return f"BEGIN TRANSACTION{this}{mark}
```

File: sqlglot/parser.py [hunk 22]
```diff
@@ -1205,7 +1205,7 @@ def extend_props(temp_props: t.Optional[exp.Properties]) -> None:
             extend_props(self._parse_properties())
 
             self._match(TokenType.ALIAS)
-            begin = self._match(TokenType.BEGIN)
+            begin = self._match_text_seq("BEGIN")
```
Comment (GeorgeSittas) [inline sqlglot/parser.py:1208]: This needs to be reverted

File: sqlglot/dialects/tsql.py (hunk 20 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Done, pushed.

File: sqlglot/dialects/tsql.py [hunk 23]
```diff
@@ -365,6 +368,57 @@ class Parser(parser.Parser):
...
+            [ ; ]
+            """
+            if self._match_texts(("TRAN", "TRANSACTION")):
+                # we have a transaction and not a BEGIN
+                transaction = self.expression(exp.Transaction, this=self._parse_id_var())
+                if self._match_text_seq("WITH", "MARK"):
+                    transaction.set("mark", self._parse_string())
+                return transaction
+
+            return self.expression(exp.Command, this="BEGIN")
```
Comment (GeorgeSittas) [inline sqlglot/dialects/tsql.py:420]: This should instead be:

```suggestion
            return self._parse_as_command(self._prev)
```

The `exp.Command` class is generally used as a fallback and just stores a raw string from the "command" token to the end of the chunk being parsed. This allows the parser to be lenient instead of failing, even in cases where we don't actually parse the syntax of a dialect-specific command.

Implementing this suggestion covers what you've already tested + more, so it's more robust.

File: sqlglot/parser.py (hunk 22 above)
Comment (dmoore247) [reply sqlglot/parser.py]: reverted.

File: sqlglot/dialects/tsql.py (hunk 21 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py:609]: I found in some cases, `this` was `TRAN` or `TRANSACTION` and ended up with `BEGIN TRANSACTION TRAN`

Comment (GeorgeSittas) [reply sqlglot/dialects/tsql.py:609]: What's an example where this happens?

File: sqlglot/dialects/tsql.py (hunk 23 above)
Comment (dmoore247) [reply sqlglot/dialects/tsql.py]: Done.

**Comment by dmoore247** [comment]
@GeorgeSittas Looks like I resolved all the issues.
